### PR TITLE
enable `rust_2018_idioms` lint group for doctests

### LINKED ISSUE #99144: Run compiler doctests in CI
Right now these are only ever run locally, which means they often break. We should run these in CI.

See https://github.com/rust-lang/rust/issues/99144#issuecomment-1269609271 for mentoring instructions.

_Originally posted by @jyn514 in https://github.com/rust-lang/rust/issues/95994#issuecomment-1097487862_

### LINKED ISSUE #106086: Enable ELIDED_LIFETIMES_IN_PATHS lint globally in rustc
We should enable https://doc.rust-lang.org/beta/nightly-rustc/rustc_lint/builtin/static.ELIDED_LIFETIMES_IN_PATHS.html globally in rustc, I would be a fan of that change to make sure we enforce `LabelText<'_>` over `LabelText`.

https://github.com/rust-lang/rust/blob/f94c926aecf226a12ffc5c7f09156390dde545ca/src/bootstrap/builder.rs#L1662-L1667 would be the place to add that.

_Originally posted by @compiler-errors in https://github.com/rust-lang/rust/pull/105975#discussion_r1055833547_

### PR DESCRIPTION
**Pull request by onur-ozkan**
With this change, `rust_2018_idioms` lint group will be enabled for compiler/libstd doctests.

Resolves #106086 
Resolves #99144

Signed-off-by: ozkanonur <work@onurozkan.dev>

### CODE REVIEW

File: src/librustdoc/doctest.rs
Comment (Nilstrieb) [inline src/librustdoc/doctest.rs:352]: This enables the lint for _everyone_ using rustdoc, which is not what we want here. You should instead add it here as described in the issue, so it's only enabled in this repository. You also don't need the UI test then.
https://github.com/rust-lang/rust/blob/f94c926aecf226a12ffc5c7f09156390dde545ca/src/bootstrap/builder.rs#L1662-L1667

Comment (ozkanonur) [reply src/librustdoc/doctest.rs:352]: > This enables the lint for everyone using rustdoc, which is not what we want here.

I wasn't sure about that. Thanks for clarification.

Comment (Nilstrieb) [reply src/librustdoc/doctest.rs:352]: Regarding my posted solution, I was wrong and misread the issue, adding it to the bootstrap builder doesn't help.

File: src/librustdoc/doctest.rs
Comment (ozkanonur) [inline src/librustdoc/doctest.rs:356]: I am unsure if this approach is correct. Please double check this block with https://github.com/rust-lang/rust/blob/91178d27b4c38f5f2768ea86c731ecaca09fa4ac/src/bootstrap/builder.rs#L1736

**Comment by ozkanonur** [comment]
Because of https://github.com/rust-lang/rust/blob/91178d27b4c38f5f2768ea86c731ecaca09fa4ac/src/bootstrap/builder.rs#L1722

this line, we get lots of errors on doctest. I can fix the doctests warnings/errors using attributes like `compile_fail` or directly update the code blocks in doctests, or remove the line above. Which one I should do? @jyn514 @Nilstrieb

**Comment by jyn514** [comment]
@ozkanonur you should fix the warnings - if it's a lot of effort, we can split it up by crate on the original issue and ask for volunteers to help fix the issues.

File: src/librustdoc/doctest.rs
Comment (jyn514) [reply src/librustdoc/doctest.rs:356]: yes, this is not correct: we should only enable it for crates in the rust-lang/rust workspace, not all rust crates anywhere in the ecosystem.

Please also use `-Zcrate-attr`, not attributes in each crate, to ensure the lint is automatically enabled when someone adds a new rustc_ crate. See https://github.com/rust-lang/rust/blob/3b1c8a94a4e8a6ba8bc7b39cc3580db9e5b72295/src/bootstrap/compile.rs#L393 for an example.

File: src/bootstrap/compile.rs
Comment (jyn514) [inline src/bootstrap/compile.rs:763]: I'm not sure this has the intended effect? I think if you just make this change and revert the changes to doctests, you won't see any warnings; the attribute is being applied to the crate itself, not to the doctests in it. See my comment above about doc(test(warn(...))).

**Review by jyn514 — COMMENTED** [review]
The compiler changes look good to me, but we should probably have someone from libs sign off on the changes to the public docs.

File: src/bootstrap/compile.rs
Comment (ozkanonur) [reply src/bootstrap/compile.rs:763]: I thought this attribute will be enough because it's under the `rustdocflag`. Also, it works as expected.

### Here is the demonstration:

Diff:
```diff
diff --git a/compiler/rustc_graphviz/src/lib.rs b/compiler/rustc_graphviz/src/lib.rs
index 5d86d895817..549cf4bf850 100644
--- a/compiler/rustc_graphviz/src/lib.rs
+++ b/compiler/rustc_graphviz/src/lib.rs
@@ -164,7 +164,7 @@
 //!     fn node_id(&'a self, n: &Nd) -> dot::Id<'a> {
 //!         dot::Id::new(format!("N{}", n)).unwrap()
 //!     }
-//!     fn node_label(&self, n: &Nd) -> dot::LabelText<'_> {
+//!     fn node_label(&self, n: &Nd) -> dot::LabelText {
 //!         dot::LabelText::LabelStr(self.nodes[*n].into())
 //!     }
 //!     fn edge_label(&self, _: &Ed<'_>) -> dot::LabelText<'_> {
@@ -177,7 +177,7 @@
 //!     type Edge = Ed<'a>;
 //!     fn nodes(&self) -> dot::Nodes<'a,Nd> { (0..self.nodes.len()).collect() }
 //!     fn edges(&'a self) -> dot::Edges<'a,Ed<'a>> { self.edges.iter().collect() }
-//!     fn source(&self, e: &Ed<'_>) -> Nd { let & &(s,_) = e; s }
+//!     fn source(&self, e: &Ed) -> Nd { let & &(s,_) = e; s }
 //!     fn target(&self, e: &Ed<'_>) -> Nd { let & &(_,t) = e; t }
 //! }
 //!
```

Part of the `x test compiler/rustc_graphviz` output:
```sh
running 6 tests
...F..
failures:

---- src/lib.rs - (line 142) stdout ----
error: hidden lifetime parameters in types are deprecated
  --> src/lib.rs:167:42
   |
28 |     fn node_label(&self, n: &Nd) -> dot::LabelText {
   |                                     -----^^^^^^^^^
   |                                     |
   |                                     expected lifetime parameter
   |
note: the lint level is defined here
  --> <crate attribute>:1:6
   |
1  | deny(elided_lifetimes_in_paths)
   |      ^^^^^^^^^^^^^^^^^^^^^^^^^
help: indicate the anonymous lifetime
   |
28 |     fn node_label(&self, n: &Nd) -> dot::LabelText<'_> {
   |                                                   ++++

error: hidden lifetime parameters in types are deprecated
  --> src/lib.rs:180:26
   |
41 |     fn source(&self, e: &Ed) -> Nd { let & &(s,_) = e; s }
   |                          ^^ expected lifetime parameter
   |
help: indicate the anonymous lifetime
   |
41 |     fn source(&self, e: &Ed<'_>) -> Nd { let & &(s,_) = e; s }
   |                            ++++

error: aborting due to 2 previous errors

Couldn't compile the test.

failures:
    src/lib.rs - (line 142)

test result: FAILED. 5 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.22s
```

**Comment by jyn514** [comment]
r? libs

Do the changes to the standard library documentation look like an improvement?

File: src/bootstrap/compile.rs
Comment (ozkanonur) [reply src/bootstrap/compile.rs:763]: > Ok, no need to change it, but maybe leave a comment in builder.rs explaining what's going on?

Sure.

What do you think this diff?
```diff
diff --git a/src/bootstrap/builder.rs b/src/bootstrap/builder.rs
index b54bf432621..34304e4854c 100644
--- a/src/bootstrap/builder.rs
+++ b/src/bootstrap/builder.rs
@@ -2048,6 +2048,10 @@ pub(crate) fn open_in_browser(&self, path: impl AsRef<Path>) {
 mod tests;

 #[derive(Debug, Clone)]
+/// Represents flag values in `String` form with whitespace delimiter to pass it to the compiler later.
+///
+/// `-Z crate-attr` flags will be applied recursively on the target code using the `rustc_parse::parser::Parser`.
+/// See `rustc_builtin_macros::cmdline_attrs::inject` for more information.
 struct Rustflags(String, TargetSelection);

 impl Rustflags {
```

Comment (jyn514) [reply src/bootstrap/compile.rs:763]: That works :) I was imagining a comment where you added crate-attr but this is good too.

Comment (ozkanonur) [reply src/bootstrap/compile.rs:763]: > maybe leave a comment in builder.rs explaining what's going on?

I guess this was a typo then, `crate-attr` is added in `compile.rs` :)

**Comment by cuviper** [comment]
The doc changes look fine to me. But maybe we should go further -- did you consider the whole `rust-2018-idioms` lint group? We already enable that for all in-tree sources in `Builder::cargo`, so I think it would make sense for docs too. I see that this group was mentioned in the issue, but it seems not knowing the applicability of `-Zcrate-attr` here.

**Comment by jyn514** [comment]
👍 enabling this for all idioms sounds like a good idea

**Comment by ozkanonur** [comment]
in `rust-2018-idioms`, `elided_lifetimes_in_paths` lint is allowed here:

https://github.com/rust-lang/rust/blob/c0b87359599629d69a7b37318088dac3a5d67c9c/compiler/rustc_lint_defs/src/builtin.rs#L1731-L1738

The main purpose of this PR was simply enabling `elided_lifetimes_in_paths` in in-tree doctests. If I turn `Allow` into `Warn` or `Deny`, this will effect the whole ecosystem(where `rust-2018-idioms` is enabled) as you already know. Please confirm if I understand it correctly then I will update the diff. @jyn514 @cuviper

**Comment by mejrs** [comment]
> If I turn Allow into Warn or Deny, this will effect the whole ecosystem(where rust-2018-idioms is enabled) as you already know. Please confirm if I understand correctly that you want me to do enable this lint as a default in rust-2018-idioms then I will update the diff.

No, if you deny on`rust-2018-idioms` , it should deny on `elided_lifetimes_in_paths` too. You shouldn't be changing the level of the lint.

**Comment by ozkanonur** [comment]
> > If I turn Allow into Warn or Deny, this will effect the whole ecosystem(where rust-2018-idioms is enabled) as you already know. Please confirm if I understand correctly that you want me to do enable this lint as a default in rust-2018-idioms then I will update the diff.
> 
> No, if you deny on`rust-2018-idioms` , it should deny on `elided_lifetimes_in_paths` too. You shouldn't be changing the level of the lint.

I was on the wrong way then, much thanks:)

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs
Comment (ozkanonur) [inline compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs:24]: This test doesn't provide `E0623`, that's why it's removed. Can't see the error code of the error. It just raises `error: lifetime may not live long enough`

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/find_anon_type.rs
Comment (ozkanonur) [inline compiler/rustc_infer/src/infer/error_reporting/nice_region_error/find_anon_type.rs:17]: Same here

File: compiler/rustc_infer/src/infer/outlives/verify.rs
Comment (ozkanonur) [inline compiler/rustc_infer/src/infer/outlives/verify.rs:298]: This can be compiled without any problem.

**Comment by ozkanonur** [comment]
PR is ready for review @jyn514

**Comment by jyn514** [comment]
This appears to now have a lot of overlap with https://github.com/rust-lang/rust/pull/106486 and https://github.com/rust-lang/rust/pull/106574.

**Comment by ozkanonur** [comment]
What do you think cherry-picking those PRs in this one?

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (Nilstrieb) [inline compiler/rustc_ast_lowering/src/lib.rs:1404]: Can't you just leave these here as-is? They shouldn't complain because `doctest = false` is set (and the other PRs jyn mentioned already are working on enabling it).

File: compiler/rustc_graphviz/src/lib.rs
Comment (Nilstrieb) [inline compiler/rustc_graphviz/src/lib.rs:42]: Can you remove it alltogether? The comment and your `allow` seem to disagree with each other.

File: compiler/rustc_graphviz/src/lib.rs
Comment (Nilstrieb) [inline compiler/rustc_graphviz/src/lib.rs:151]: Same here.

File: compiler/rustc_graphviz/src/lib.rs
Comment (Nilstrieb) [inline compiler/rustc_graphviz/src/lib.rs:217]: And here.

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs
Comment (Nilstrieb) [inline compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:88]: This doesn't look like pseudo-Rust to me and doesn't need an ignore.

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs
Comment (Nilstrieb) [inline compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:97]: If you move the "and an impl" comment above you can merge the code blocks and also remove the ignore from here (you'll need to declare a  `struct Foo;`)

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs
Comment (Nilstrieb) [inline compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:1096]: This should be `compile_fail`

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (Nilstrieb) [inline compiler/rustc_lint_defs/src/builtin.rs:2140]: Why do we need this here? Does this code not trigger `explicit_outlives_requirements` (that would be pretty bad lol)?

**Review by Noratrieb — CHANGES_REQUESTED** [review]
Most changes look good except for a few where I left a comment.

Also, can you remove all the changes that also appear in #106574 or #106486. This PR should still work even without it.

File: compiler/rustc_graphviz/src/lib.rs
Comment (ozkanonur) [reply compiler/rustc_graphviz/src/lib.rs:42]: I had to implement that lines to avoid following error:

```text
failures:

---- src/lib.rs - (line 37) stdout ----
error: `extern crate` is not idiomatic in the new edition
 --> src/lib.rs:39:1
  |
5 | extern crate r#rustc_graphviz;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: convert it to a `use`
  |
note: the lint level is defined here
 --> src/lib.rs:36:9
  |
2 | #![deny(warnings)]
  |         ^^^^^^^^
  = note: `#[deny(unused_extern_crates)]` implied by `#[deny(warnings)]`

error: aborting due to previous error

Couldn't compile the test.

failures:
    src/lib.rs - (line 37)
```

Even the allow itself doesn't help, following extern implementation has to be there for fix. That's the workaround I found, not sure if there is a better way.

File: compiler/rustc_graphviz/src/lib.rs
Comment (ozkanonur) [reply compiler/rustc_graphviz/src/lib.rs:151]: https://github.com/rust-lang/rust/pull/106621#discussion_r1080574124

File: compiler/rustc_graphviz/src/lib.rs
Comment (ozkanonur) [reply compiler/rustc_graphviz/src/lib.rs:217]: https://github.com/rust-lang/rust/pull/106621#discussion_r1080574124

File: compiler/rustc_graphviz/src/lib.rs
Comment (jyn514) [reply compiler/rustc_graphviz/src/lib.rs:42]: oh ugh, the `r#rustc_graphviz` is injected by rustdoc itself, I don't think there's a way to avoid the lint triggering. Maybe rustdoc could inject the allow for just the `extern crate` statement so it doesn't affect any other part of the code?

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (ozkanonur) [reply compiler/rustc_lint_defs/src/builtin.rs:2140]: It triggers it but as a `warn`, so I had to deny warnings in order to fail the compilation of it.

here is the stdout when there is no `#![deny(warnings)]`:
```text
---- src/builtin.rs - builtin::EXPLICIT_OUTLIVES_REQUIREMENTS (line 2136) stdout ----
warning: outlives requirements can be inferred
 --> src/builtin.rs:2140:24
  |
6 |   struct SharedRef<'a, T>
  |  ________________________^
7 | | where
8 | |     T: 'a,
  | |__________^ help: remove this bound
  |
note: the lint level is defined here
 --> <crate attribute>:1:6
  |
1 | warn(rust_2018_idioms)
  |      ^^^^^^^^^^^^^^^^
  = note: `#[warn(explicit_outlives_requirements)]` implied by `#[warn(rust_2018_idioms)]`

warning: 1 warning emitted

Test compiled successfully, but it's marked `compile_fail`.
```

File: compiler/rustc_graphviz/src/lib.rs
Comment (ozkanonur) [reply compiler/rustc_graphviz/src/lib.rs:42]: > oh ugh, the `r#rustc_graphviz` is injected by rustdoc itself, I don't think there's a way to avoid the lint triggering. Maybe rustdoc could inject the allow for just the `extern crate` statement so it doesn't affect any other part of the code?

Hmm, sounds like better solution.

```diff
diff --git a/src/librustdoc/doctest.rs b/src/librustdoc/doctest.rs
index 81d9c46447a..7f9ac790d1d 100644
--- a/src/librustdoc/doctest.rs
+++ b/src/librustdoc/doctest.rs
@@ -682,6 +682,8 @@ pub(crate) fn make_test(
             // parse the source, but only has false positives, not false
             // negatives.
             if s.contains(crate_name) {
+                // Allow `unused_extern_crates` lint rule to avoid lint errors/warnings
+                prog.push_str(&format!("#[allow(unused_extern_crates)]\n"));
                 prog.push_str(&format!("extern crate r#{crate_name};\n"));
                 line_offset += 1;
             }
```

This handles the issue

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (ozkanonur) [reply compiler/rustc_ast_lowering/src/lib.rs:1404]: It does fail actually.

stdout:
```text
---- src/lib.rs - LoweringContext<'a,'hir>::lower_opaque_impl_trait (line 1404) stdout ----
error[E0404]: expected trait, found derive macro `Debug`
 --> src/lib.rs:1405:16
  |
3 | fn test<'a, T: Debug>(x: &'a T) -> impl Debug + 'a {
  |                ^^^^^ not a trait
  |
help: consider importing one of these items instead
  |
2 | use core::fmt::Debug;
  |
2 | use std::fmt::Debug;
  |

error[E0404]: expected trait, found derive macro `Debug`
 --> src/lib.rs:1405:41
  |
3 | fn test<'a, T: Debug>(x: &'a T) -> impl Debug + 'a {
  |                                         ^^^^^ not a trait
  |
help: consider importing one of these items instead
  |
2 | use core::fmt::Debug;
  |
2 | use std::fmt::Debug;
```

since this fits better, I updated it as:
```diff
diff --git a/compiler/rustc_ast_lowering/src/lib.rs b/compiler/rustc_ast_lowering/src/lib.rs
index 14f1a0f4019..5cb681b930b 100644
--- a/compiler/rustc_ast_lowering/src/lib.rs
+++ b/compiler/rustc_ast_lowering/src/lib.rs
@@ -1401,7 +1401,9 @@ fn lower_ty_direct(&mut self, t: &Ty, itctx: &ImplTraitContext) -> hir::Ty<'hir>
     ///
     /// Given a function definition like:
     ///
-    /// ```rust,ignore (pseudo-Rust)
+    /// ```rust
+    /// use std::fmt::Debug;
+    ///
     /// fn test<'a, T: Debug>(x: &'a T) -> impl Debug + 'a {
     ///     x
     /// }
```

**Comment by ozkanonur** [comment]
I fixed the review notes which I add an emoji(:thumbsup:)

r? @Nilstrieb

**Comment by Nilstrieb** [comment]
I don't understand why crates with disabled doctests `rustc_ast_infer` is suddenly giving errors.. do you know why?

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs
Comment (Nilstrieb) [reply compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs:24]: Not all errors have codes, that's fine.

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (Nilstrieb) [reply compiler/rustc_lint_defs/src/builtin.rs:2140]: ugh, does the warn override the deny :(

**Comment by jyn514** [comment]
> I don't understand why crates with disabled doctests `rustc_ast_infer` is suddenly giving errors.. do you know why?

`doctest = false` only affects whether the tests are run by default, not whether they're run when you explicitly do `x test rustc_ast_infer` 🙃

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs
Comment (ozkanonur) [reply compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:998]: It can't compile.

```text
failures:

---- src/deriving/generic/mod.rs - deriving::generic::MethodDef<'a>::expand_struct_method_body (line 998) stdout ----
error: expected expression, found `==`
 --> src/deriving/generic/mod.rs:1003:20
  |
7 |         { self.x } == { other.y } && { self.y } == { other.y }
  |                    ^^ expected expression

error[E0404]: expected trait, found derive macro `Hash`
  --> src/deriving/generic/mod.rs:1006:6
   |
10 | impl Hash for A {
   |      ^^^^ not a trait
   |
help: consider importing one of these items instead
   |
2  | use core::hash::Hash;
   |
2  | use std::hash::Hash;
   |

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0404`.
Couldn't compile the test.
```

I didn't wanted to change the code, so I preferred to ignore it. I guess updating the code is better in this case.

In order to fix it, this diff has to be applied:
```diff
+++ b/compiler/rustc_builtin_macros/src/deriving/generic/mod.rs
@@ -995,12 +995,14 @@ fn create_method(
     /// `&self.x` because that might cause an unaligned ref. So for any trait
     /// method that takes a reference, if the struct impls `Copy` then we use a
     /// local block to force a copy:
-    /// ```rust,ignore (pseudo-Rust)
+    /// ```rust
+    /// use std::hash::Hash;
+    ///
     /// # struct A { x: u8, y: u8 }
     /// impl PartialEq for A {
     ///     fn eq(&self, other: &A) -> bool {
     ///         // Desugars to `{ self.x }.eq(&{ other.y }) && ...`
-    ///         { self.x } == { other.y } && { self.y } == { other.y }
+    ///         self.x == { other.y } && { self.y } == { other.y }
     ///     }
     /// }
     /// impl Hash for A {
```

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (ozkanonur) [reply compiler/rustc_lint_defs/src/builtin.rs:2140]: If it's not a bug, it does override :)

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs
Comment (Nilstrieb) [inline compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1005]: It actually expands to the left code though.
```rust
#[derive(PartialEq, Clone, Copy)]
#[repr(packed)]
struct A { x: u8, y: u8 }
```
```rust
impl ::core::cmp::PartialEq for A {
    #[inline]
    fn eq(&self, other: &A) -> bool {
        { self.x } == { other.x } && { self.y } == { other.y }
    }
}
```
So you should `ignore` it instead I guess

File: compiler/rustc_infer/src/infer/outlives/verify.rs
Comment (Nilstrieb) [reply compiler/rustc_infer/src/infer/outlives/verify.rs:292]: The sentence below is still wrong and I'd like to get confirmation from people knowing this code better that this is indeed good and correct and whether the docs should be adjusted further

**Comment by Nilstrieb** [comment]
This also duplicates everything from #106523, I'm not entirely sure about the best way to handle this.

**Comment by ozkanonur** [comment]
> The sentence below is still wrong and I'd like to get confirmation from people knowing this code better that this is indeed good and correct and whether the docs should be adjusted further

Could be better.

> This also duplicates everything from #106523, I'm not entirely sure about the best way to handle this.

I think cherry-picking those commits into this PR could be simplest way to deal with. Or we can wait until they get merged(which will block this PR for pretty long time)

**Comment by ozkanonur** [comment]
#106523 seems inactive 

@rustbot ready

File: library/std/src/primitive_docs.rs
Comment (Mark-Simulacrum) [inline library/std/src/primitive_docs.rs:558]: This doesn't feel like a reasonable fix to me? In particular if `extern crate libc;` *is* unused, shouldn't it just be removed?

File: src/bootstrap/builder.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/builder.rs:2150]: Typically doc-comments are placed before a derive(...) to keep that close to the struct in question, I find that much clearer.

File: src/librustdoc/doctest.rs
Comment (ozkanonur) [reply src/librustdoc/doctest.rs:683]: Yeah I agree, will do it

File: library/std/src/primitive_docs.rs
Comment (ozkanonur) [reply library/std/src/primitive_docs.rs:558]: It's not used in `library/std/src/primitive_docs` but used in `library/core/src/primitive_docs`.

So I have to put this on `primitive_docs` under both `core` and `std` to overcome tidy and doctest errors.

**Comment by ozkanonur** [comment]
no more blocked with #110798

@rustbot ready

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (GuillaumeGomez) [inline compiler/rustc_ast_lowering/src/lib.rs:1486]: If it was compiling before, why not add a hidden import instead?

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (GuillaumeGomez) [inline compiler/rustc_ast_lowering/src/lib.rs:1492]: If it was compiling before, why not add a hidden import instead?

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs
Comment (GuillaumeGomez) [inline compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1041]: Why ignoring this one?

Comment (ozkanonur) [reply compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1041]: https://github.com/rust-lang/rust/pull/106621#discussion_r1085864118

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (ozkanonur) [reply compiler/rustc_ast_lowering/src/lib.rs:1492]: It's using dummy type defined on the previous doc-comment which is also ignored.

File: compiler/rustc_ast_lowering/src/lib.rs
Comment (ozkanonur) [reply compiler/rustc_ast_lowering/src/lib.rs:1486]: > why not add a hidden import instead?

There are multiple errors, and one of them was:
```
error: unconstrained opaque type
 --> src/lib.rs:1488:30
  |
4 | type TestReturn<'a, 'x, T> = impl Debug + 'x;
  |                              ^^^^^^^^^^^^^^^
  |
  = note: `TestReturn` must be used in combination with a concrete type within the same item
```

File: compiler/rustc_infer/src/infer/outlives/verify.rs
Comment (Mark-Simulacrum) [reply compiler/rustc_infer/src/infer/outlives/verify.rs:287]: Can we switch this to an ignore test and add a comment referencing that this does compile today? I think that would be better than deleting this test and allow us to move forward with this PR.

**Pull request merged**
